json.go

```diff
@@ -0,0 +1,186 @@
+package matrixfederation
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"sort"
+	"unicode/utf8"
+)
+
+// CanonicalJSON re-encodes the JSON in a cannonical encoding. The encoding is
+// the shortest possible encoding using integer values with sorted object keys.
+func CanonicalJSON(input []byte) ([]byte, error) {
+	sorted, err := SortJSON(input, make([]byte, 0, len(input)))
+	if err != nil {
+		return nil, err
+	}
+	return CompactJSON(sorted, make([]byte, 0, len(sorted))), nil
+}
+
+// SortJSON reencodes the JSON with the object keys sorted by lexicographically
+// by codepoint.
+func SortJSON(input, output []byte) ([]byte, error) {
+	// Skip to the first character that isn't whitespace.
+	var i int
+	firstChar := input[i]
+	for firstChar <= ' ' {
+		firstChar = input[i]
+		i++
+	}
+	if firstChar == '[' {
+		return sortJSONArray(input, output)
+	}
+	if firstChar == '{' {
+		return sortJSONObject(input, output)
+	}
+	return append(output, input...), nil
```

Comment:
You should make it clear in the docs that if you don't give `SortJSON` valid JSON it will just spit back the input you gave *rather than return an error*.